flink-runtime/src/test/java/org/apache/flink/runtime/rest/AbstractHandlerTest.java

```diff
@@ -0,0 +1,190 @@
+import org.apache.flink.runtime.rest.messages.UntypedResponseMessageHeaders;
+import org.apache.flink.runtime.rpc.RpcUtils;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.TestingRestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.Unpooled;
+import org.apache.flink.shaded.netty4.io.netty.channel.Channel;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.DefaultFullHttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpMethod;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpVersion;
+import org.apache.flink.shaded.netty4.io.netty.util.Attribute;
+import org.apache.flink.shaded.netty4.io.netty.util.AttributeKey;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import javax.annotation.Nonnull;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link AbstractHandler}.
+ */
+public class AbstractHandlerTest {
+
+	@Rule
+	public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	@Test
+	public void testFileCleanup() throws Exception {
+		final Path file = temporaryFolder.newFile().toPath();
+
+		final String restAddress = "http://localhost:1234";
+		RestfulGateway mockRestfulGateway = TestingRestfulGateway.newBuilder()
+			.setRestAddress(restAddress)
+			.build();
+
+		final GatewayRetriever<RestfulGateway> mockGatewayRetriever = () ->
+			CompletableFuture.completedFuture(mockRestfulGateway);
+
+		TestHandler handler = new TestHandler(CompletableFuture.completedFuture(restAddress), mockGatewayRetriever);
+
+		RouteResult<?> routeResult = new RouteResult<>("", "", Collections.emptyMap(), Collections.emptyMap(), "");
+		HttpRequest request = new DefaultFullHttpRequest(
+			HttpVersion.HTTP_1_1,
+			HttpMethod.GET,
+			TestHandler.TestHeaders.INSTANCE.getTargetRestEndpointURL(),
+			Unpooled.wrappedBuffer(new byte[0]));
+		RoutedRequest<?> routerRequest = new RoutedRequest<>(routeResult, request);
+
+		Attribute<FileUploads> attribute = new SimpleAttribute();
+		attribute.set(new FileUploads(Collections.emptyList(), Collections.singleton(file)));
+		Channel channel = mock(Channel.class);
+		when(channel.attr(any(AttributeKey.class))).thenReturn(attribute);
+
+		ChannelHandlerContext context = mock(ChannelHandlerContext.class);
+		when(context.channel()).thenReturn(channel);
```

Comment:
Could we use the `TestingChannelHandlerContext` instead of mocking these interfaces with Mockito?

Reply:
Are you referring to the one in `AbstractTaskManagerFileHandlerTest`? If so, I don't really see a difference between mocking 1 method and having an "implementation" that only really implements a miniscule subset.

I initially implemented both the `Channel` and `ChannelHandlerContext` with the same behavior as the mocks, but all that was was add 50+ methods that are just noise.

Reply:
Yes, I guess that one would have to pull it out and create a builder similar to `TestingRestfulGateway`. One would not have to add a build method for all methods but one could incrementally extend this once it is needed. But this might be a bit too much work given that `ChannelHandlerContext` is hardly used by our tests.